# Объявление вида ошибки

Иногда для упрощения рукописи необходимо скрыть все виды ошибок за какой-то одной ошибкой. Мы скроем их за пользовательской ошибкой.

Ржачина позволяет нам определить наш собственный вид ошибок. 
В общем случае "хороший" вид ошибки должен:

- Представлять разные ошибки с таким же видом
- Предоставлять хорошее сообщение об ошибке пользователю
- Легко сравниваться с другими видами
    - Хорошо: `Err(EmptyVec)`
    - Плохо: `Err("Пожалуйста, используйте вектор хотя бы с одним элементом".to_owned())`
- Содержать сведения об ошибке
    - Хорошо: `Err(BadChar(c, position))`
    - Плохо: `Err("+ не может быть использован в данном месте".to_owned())`
- Хорошо сочетаться с другими ошибками

```rust,editable
use std::error;
use std::fmt;

type Result<T> = std::result::Result<T, DoubleError>;

// Определите виды ошибок. Они могут быть настроены для наших случаев обработки ошибок.
// Теперь мы сможем написать наши собственные ошибки, реализовать приведение до основной ошибки
// или сделать что-то ещё между приведениями.
#[derive(Debug, Clone)]
struct DoubleError;

// Создание ошибки полностью отделена от того, как она отображается.
// Нет необходимости в загромождении сложной логикой построения отображения ошибки.
//
// Мы не храним дополнительной сведений об ошибках. Это означает, что мы не можем вывести строку, которую не удалось обработать, без изменения наших видов.
impl fmt::Display for DoubleError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "неверный первый элемент")
    }
}

// Все ошибки сворачиваются в одну.
impl error::Error for DoubleError {
    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
        // Общая ошибка не обрабатывается.
        None
    }
}

fn double_first(vec: Vec<&str>) -> Result<i32> {
    vec.first()
        // Изменим ошибку на наш новый вид.
        .ok_or(DoubleError)
        .and_then(|s| {
            s.parse::<i32>()
                // Обновим вид ошибки также здесь.
                .map_err(|_| DoubleError)
                .map(|i| 2 * i)
        })
}

fn print(result: Result<i32>) {
    match result {
        Ok(n) => println!("Первое удвоение {}", n),
        Err(e) => println!("Ошибка: {}", e),
    }
}

fn main() {
    let numbers = vec!["42", "93", "18"];
    let empty = vec![];
    let strings = vec!["tofu", "93", "18"];

    print(double_first(numbers));
    print(double_first(empty));
    print(double_first(strings));
}
```
